### PR TITLE
Buildkite CI

### PR DESCRIPTION
**Pull request by claireimprobable**
#### Description
This PR adds BuildKite capabilities for the Unreal GDK. The pipeline is set up to automatically build changes to a branch when pushed. This is intended to be a first functional set up, and there are ongoing efforts to optimise build times and improve the build scripts. 

#### Tests
UnrealGDK changes are automatically built in BuildKite and are green. 

#### Documentation
For general BuildKite documentation when modifying this pipeline, see brevi.link/buildkite-docs

#### Primary reviewers
@m-samiec 
@DoomGerbil

### CODE REVIEW

File: .buildkite/premerge.definition.yaml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+agent_queue_id: trigger-pipelines
+description: Build Unreal GDK
+github:
+  branch_configuration:
```
Comment (DoomGerbil) [inline .buildkite/premerge.definition.yaml]: This will need to change before merging, since it restricts CI such that it will only run against this branch.  I think you probably want to set this to `branch_configuration: []` and remove the specified branch to make it run on all branches.
```suggestion
  branch_configuration: []
  
```

File: .buildkite/premerge.steps.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+---
```
Comment (DoomGerbil) [inline .buildkite/premerge.steps.yaml]: I tweaked this file to remove a bunch of boilerplate comments and change the step label to something that works a bit better with the GitHub status checks API (spaces in status names can make it unhappy sometimes).

File: ci/setup-and-build.ps1 [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    )
+
+    # Start this tracing span.
+    Start-Process -NoNewWindow "imp-ci" -ArgumentList @(`
+        "events", "new", `
+        "--name", "$($event_name)", `
+        "--child-of", "$($event_parent)"
+    ) | Out-Null
+
+    Write-Log "--- $($event_name)"
```
Comment (DoomGerbil) [inline ci/setup-and-build.ps1]: Remove the leading `---` from here, since it's already pre-pended by the `Write-Log` function (thus this makes things log as `--- --- Unreal Engine` instead of the intended `--- Unreal Engine`.  

Sorry about that, I wrote this bug originally.  :)
```suggestion
    Write-Log "$($event_name)"
```

File: ci/setup-and-build.ps1 [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        }
+    popd
+    Finish-Event "download-unreal-engine" "build-unreal-gdk-:windows:"
+
+
+    $unreal_path = "$($gdk_home)\UnrealEngine"
+    Write-Log "Setting UNREAL_HOME environment variable to $($unreal_path)"
+    [Environment]::SetEnvironmentVariable("UNREAL_HOME", "$($unreal_path)", "Machine")
+
+    ## THIS REPLACES THE OLD SETUP.BAT SCRIPT
```
Comment (DoomGerbil) [inline ci/setup-and-build.ps1]: I don't know if this is relevant, but maybe worth considering if replacing existing setup scripts intended to be run manually - if you want this script to be usable both inside and outside of CI, check for the environment variable `BUILDKITE`.  

This will be set to `true` if you're running in CI, and unset if you're not.

File: ci/setup-and-build.ps1 [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    Expand-Archive -Path "$($core_sdk_dir)\worker_sdk\c-dynamic-x86_64-msvc_md-win32.zip" -DestinationPath "$($binaries_dir)\Win64\" -Force
+    Expand-Archive -Path "$($core_sdk_dir)\worker_sdk\c-dynamic-x86_64-gcc_libstdcpp-linux.zip" -DestinationPath "$($binaries_dir)\Linux\" -Force
+    Finish-Event "extract-spatial-packages" "build-unreal-gdk-:windows:"
+
+
+    # Copy from binaries_dir
+    Copy-Item "$($binaries_dir)\Win64\include\*" "$($worker_sdk_dir)\" -Force -Recurse
+
+    Write-Log "Fetch MSBUILD_EXE location"
+    $msbuild_exe = "${env:ProgramFiles(x86)}\MSBuild\14.0\bin\MSBuild.exe"
```
Comment (DoomGerbil) [inline ci/setup-and-build.ps1]: You could set the location of MSBuild.exe as a `param` on the script that defaults to this value.

That way, if someone has a custom installation location, they can override it on the command line and it will work.

File: ci/unreal-engine.version [hunk 6]
```diff
@@ -0,0 +1 @@
+UnrealEngine-something
```
Comment (DoomGerbil) [inline ci/unreal-engine.version]: Just to point out the obvious, this is a placeholder until the Engine publishing pipeline is fully up and running.  Don't know if we want to make sure to fix this before merging - can probably fix it afterwards if we want to confirm that this is working for a bit before fully relying on it.

File: ci/setup-and-build.ps1 [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+
+
+    Start-Event "build-utilities" "build-unreal-gdk-:windows:"
+    #%MSBUILD_EXE% /nologo /verbosity:minimal .\SpatialGDK\Build\Programs\Improbable.Unreal.Scripts\Improbable.Unreal.Scripts.sln /property:Configuration=Release
+    $msbuild_proc = Start-Process -Wait -PassThru -NoNewWindow -FilePath "$($msbuild_exe)" -ArgumentList @(`
```
Comment (DoomGerbil) [inline ci/setup-and-build.ps1]: From looking at https://buildkite.com/improbable/unrealgdk-premerge/builds/36#65d3be64-7446-4b80-8a17-19ace05046c8/200 it seems like this is doing the same thing we were having problems with in the UE build:

1. Start a quick msbuild process
1. After the build completes, one of the MS compilers stays running for 10 minutes, just in case you want to use it again soon.
1. Powershell's `Start-Process` with `-Wait` will block until the command you supply, plus all child processes _including that MS compiler that hangs around for 10 mins_ have exited, thus blocking the script for 10 mins for no good reason.

I'll take a look, but IIRC what we ended up doing to get around this was:
1. Use `Start-Process` without `-Wait`
1. Immediately call `Wait-Process -Id (Get-Process -InputObject $process_obj).id`, which will wait for _only_ the named process, and does not care about its child processes.

That way, once your actual MSBuild invocation is done, the script should immediately move on and ignore the stupid MS compiler trying to be clever.

Comment (DoomGerbil) [reply ci/setup-and-build.ps1]: The same build run shows the same problem on the `RunUAT.bat` invocation below that actually builds the GDK.

If we can fix the unnecessary waiting, it'll cut the total time for this pipeline from 29 mins down to 9.

File: .buildkite/premerge.steps.yaml [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  # connection. Agent exits with -1 in this case.
+  exit_status: -1
+  limit: 3
+  
+common: &common
+  agents:
+    - "capable_of_building=gdk-for-unreal"
+    - "environment=production"
+    - "permission_set=pusher"
+    - "queue=v2-1549893371-183fd9b7177e2649-------z"
```
Comment (m-samiec) [inline .buildkite/premerge.steps.yaml]: Does this queue have any versioning info against it? Is it easy to validate what changes this image has?

File: SpatialGDK/Extras/core-sdk.version [hunk 9]
```diff
@@ -1 +1 @@
-13.5.0
+13.5.0
```
Comment (m-samiec) [inline SpatialGDK/Extras/core-sdk.version]: Please revert this change :)

File: ci/setup-and-build.ps1 [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+    popd
+
+    ## Create an UnrealEngine directory if it doesn't already exist
+    New-Item -Name "UnrealEngine" -ItemType Directory -Force
+
+
+    Start-Event "download-unreal-engine" "build-unreal-gdk-:windows:"
+    pushd "UnrealEngine"
+        Write-Log "Downloading the Unreal Engine artifacts from GCS"
+        $gcs_unreal_location = "$($unreal_version).zip"
```
Comment (m-samiec) [inline ci/setup-and-build.ps1]: Don't these zip files also have a timestamp in their filename?

File: ci/setup-and-build.ps1 (hunk 4 above)
Comment (m-samiec) [reply ci/setup-and-build.ps1]: I think this is a good idea, we should look to consolidate the setup process into a single script so we don't diverge functionality.

Comment (m-samiec) [reply ci/setup-and-build.ps1]: Added a follow-on task in UNR-420.

File: ci/unreal-engine.version (hunk 6 above)
Comment (m-samiec) [reply ci/unreal-engine.version]: How are we able to test this currently with this value?

Comment (DoomGerbil) [reply ci/unreal-engine.version]: @claireimprobable and I manually moved an unreal engine build to that location in the GCS bucket while we were working on the UE build pipeline.

File: .buildkite/premerge.steps.yaml (hunk 8 above)
Comment (DoomGerbil) [reply .buildkite/premerge.steps.yaml]: Yeah, the format is `v2-<timestamp>-<git_hash>-------z` (where `-------z` is a set of state markers that denote the local state of the repo at the time the build was run.

You can always look at <git_hash> of the platform repo to confirm the exact configuration that was used to build that agent, and the timestamp can be useful for determining the historical state of any external dependencies that can't be pinned (though obviously, using a specific pinned version of your dependencies is vastly preferable for proving/reconstructing state).

Comment (m-samiec) [reply .buildkite/premerge.steps.yaml]: Is it worth adding a comment that briefly summarises that?

Comment (DoomGerbil) [reply .buildkite/premerge.steps.yaml]: We probably ought to have a page outlining this on http://brevi.link/buildkite-docs.  It hasn't been needed in practice outside of the EV team so far, but it's valuable to know that you _can_ do it.  

I've filed https://improbableio.atlassian.net/browse/ENG-1319 for us to add that to the documentation.

File: ci/setup-and-build.ps1 (hunk 7 above)
Comment (padhansell) [reply ci/setup-and-build.ps1]: This is fixed now

File: ci/setup-and-build.ps1 (hunk 10 above)
Comment (padhansell) [reply ci/setup-and-build.ps1]: The timestamp has been removed from the zip name now - it's just based on the commit hash

File: ci/build-gdk.ps1 [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+pushd "$($gdk_home)"
+
+    Start-Event "build-unreal-gdk-windows" "build-unreal-gdk-:windows:"
+    pushd "SpatialGDK"
+        $win_build_proc = Start-Process -PassThru -NoNewWindow -FilePath "$($gdk_home)\UnrealEngine\Engine\Build\BatchFiles\RunUAT.bat" -ArgumentList @(`
+            "BuildPlugin", `
+            " -Plugin=`"$($gdk_home)/SpatialGDK/SpatialGDK.uplugin`"", `
+            "-TargetPlatforms=Win64", `
+            "-Package=`"$gdk_home/SpatialGDK/Intermediate/BuildPackage/Win64`"" `
```
Comment (m-samiec) [inline ci/build-gdk.ps1]:
```suggestion
            "-Package=`"$($gdk_home)/SpatialGDK/Intermediate/BuildPackage/Win64`"" `
```

File: ci/get-engine.ps1 [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+            Write-Log "Using engine version found in unreal-engine.version file: $($unreal_version)"
+        }
+    popd
+
+    ## Create an UnrealEngine directory if it doesn't already exist
+    New-Item -Name "UnrealEngine" -ItemType Directory -Force
+
+    Start-Event "download-unreal-engine" "build-unreal-gdk-:windows:"
+    pushd "UnrealEngine"
+        Write-Log "Downloading the Unreal Engine artifacts from GCS"
```
Comment (m-samiec) [inline ci/get-engine.ps1]: Perhaps move this log till after the gcs_unreal_location definiton and include it so we know what file it was looking for if it failed.

File: ci/setup-gdk.ps1 [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+    Copy-Item "$($binaries_dir)\Win64\include\*" "$($worker_sdk_dir)\" -Force -Recurse
+
+    Start-Event "build-utilities" "build-unreal-gdk-:windows:"
+    $msbuild_proc = Start-Process -PassThru -NoNewWindow -FilePath "$($msbuild_exe)" -ArgumentList @(`
+        "/nologo", `
+        "SpatialGDK\Build\Programs\Improbable.Unreal.Scripts\Improbable.Unreal.Scripts.sln", `
+        "/property:Configuration=Release" `
+    )
+
+    # Working around a powershell bug
```
Comment (m-samiec) [inline ci/setup-gdk.ps1]: Was there a jira ticket for this? If so please link the number

Comment (padhansell) [reply ci/setup-gdk.ps1]: This actually doesn't seem to be a bug per se - I think it's expected behaviour with PS. Have updated the comment to reflect this better
